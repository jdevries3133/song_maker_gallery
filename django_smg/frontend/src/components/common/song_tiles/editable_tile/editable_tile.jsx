import React, { useState } from "react";
import PropTypes from "prop-types";

import styled, {
  Label,
  Input,
  Button,
  Textarea,
  Description as DefaultDescription,
  P,
  H3,
} from "Styles";

const VALIDATION_REGEX = /^http(s)?:\/\/musiclab.chromeexperiments.com\/Song-Maker\/song\/\d{16}$/;

const Description = styled(DefaultDescription)`
  margin: 0;
  padding: 1.5vw;
`;

export const EditableTile = ({ name = "", link = "" }) => {
  const [_name, setName] = useState(name);
  const [_link, setLink] = useState(link);

  const isLinkValid = !VALIDATION_REGEX.test(_link.trim());
  const changesMade = name !== _name || link !== _link;

  const submit = (e) => {
    e.preventDefault();
  };

  return (
    <Description as="form" onSubmit={submit}>
      <Label htmlFor="name">Name</Label>
      <Input
        id="name"
        data-testid="nameInput"
        value={_name}
        onChange={(e) => setName(e.target.value)}
      />
      <div>
        <label htmlFor="link">
          <H3>Link</H3>
          {isLinkValid && <P warn>Link is not valid</P>}
        </label>
        <Textarea
          warn={isLinkValid}
          id="link"
          data-testid="linkInput"
          onChange={(e) => setLink(e.target.value)}
          value={_link}
        />
      </div>
      {changesMade && (
        <Button data-testid="submit" as="input" type="submit" value="Save" />
      )}
    </Description>
  );
};

EditableTile.propTypes = {
  name: PropTypes.string,
  link: PropTypes.string,
};
