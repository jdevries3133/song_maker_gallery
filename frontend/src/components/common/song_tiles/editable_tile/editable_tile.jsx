import React, { useState } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import axios from "axios";

import styled, {
  Label,
  Input,
  Button,
  Textarea,
  Description as DefaultDescription,
  P,
  H3,
} from "Styles";

import { LiveUpdatingTile } from "../live_updating_tile";

const VALIDATION_REGEX = /^http(s)?:\/\/musiclab.chromeexperiments.com\/Song-Maker\/song\/\d{16}$/;

const Description = styled(DefaultDescription)`
  margin: 0;
  padding: 1.5vw;
`;

const BtnContainer = styled.div`
  display: flex;
  align-item: center;
  justify-content: center;
  margin: auto;
  flex-direction: row;
`;

const updateSong = async (song, token, errCallback) => {
  axios.defaults.xsrfCookieName = "csrftoken";
  axios.defaults.xsrfHeaderName = "X-CSRFTOKEN";

  try {
    /**
     * Since the link has been updated, but the rest of the song resource has
     * not, we need to get a new song resource for the given songId to make
     * sure that the MIDI data and everything else is in sync with the songId
     */
    const res = await axios.post("/api/gallery/song_data/", {
      student_name: song.student_name,
      songId: song.songId,
    });

    /*
     * res.data has valid google data
     * res.data does not have relationships:
     *
     * - owner
     * - Gallery
     * - SongGroup
     *
     * but, `song` does have these relationships, so we can just copy them
     * over before issuing the patch request
     */
    const newSong = res.data;
    const COPY_FIELDS = ["owner", "order", "group"];
    COPY_FIELDS.forEach((field) => {
      newSong[field] = song[field];
    });

    /* Now, we can patch in the updated song */
    await axios.patch(`/api/gallery/song/${song.pk}/`, res.data, {
      headers: { Authorization: `Token ${token}` },
    });
  } catch (_) {
    errCallback();
  }
};

const deleteSong = async (pk, token, errCallback) => {
  axios.defaults.xsrfCookieName = "csrftoken";
  axios.defaults.xsrfHeaderName = "X-CSRFTOKEN";

  try {
    await axios.delete(`/api/gallery/song/${pk}/`, {
      headers: { Authorization: `Token ${token}` },
    });
  } catch (_) {
    errCallback();
  }
};

/**
 * Behavior is dynamic depending on whether the onSave prop was provided,
 * or the initialSong prop. If onSave is provided, the onSave function is
 * called with the student name and songId when the save button is pressed.
 * If the initialSong object was provided, we take that as the full cannonical
 * `Song` resource, and this component can be responsible for updating it,
 * since the pk should be included.
 */
export const _EditableTile = ({
  name = "",
  link = "",
  initialSong,
  onSave,
  token,
}) => {
  const [_name, setName] = useState(name);
  const [_link, setLink] = useState(link);
  const [errMsg, setErrMsg] = useState(null);
  const [isDeleted, setIsDeleted] = useState(false);
  // indicates successful update
  const [success, setSuccess] = useState(false);

  const isNameValid = !!_name;
  const isLinkValid = VALIDATION_REGEX.test(_link.trim());
  const changesMade = name !== _name || link !== _link;
  const songId = isLinkValid ? _link.slice(-16) : null;

  /**
   * update the song
   */
  const submit = (e) => {
    e.preventDefault();
    if (!isLinkValid) return;

    if (onSave) {
      onSave(_name, songId);
      return;
    }

    if (initialSong) {
      const newSong = {
        ...initialSong,
        songId,
        student_name: _name,
      };
      updateSong(newSong, token, () => {
        setErrMsg("Update failed due to network error. Please try again.");
      }).then(() => setSuccess(true));
    }
  };

  const deleteSelf = (e) => {
    e.preventDefault();
    deleteSong(initialSong.pk, token, () => {
      setErrMsg("Deletion failed due to network error. Please try again.");
    }).then(() => {
      setIsDeleted(true);
    });
  };

  const submitEnabled = changesMade && isLinkValid && isNameValid;

  /**
   * presence of success is temporary
   */
  if (success) {
    setTimeout(() => {
      setSuccess(false);
    }, 2000);
  }

  return (
    <Description as="form" onSubmit={submit}>
      {isDeleted ? (
        <h1>{_name}'s Song was Deleted</h1>
      ) : (
        <>
          <LiveUpdatingTile
            width={200}
            initialSong={initialSong}
            songId={isLinkValid ? songId : null}
          />
          {errMsg && <P warn>{errMsg}</P>}
          <Label htmlFor="name">Name</Label>
          {changesMade && !isNameValid && <P warn>Please enter your name.</P>}
          <Input
            id="name"
            data-testid="nameInput"
            value={_name}
            onChange={(e) => setName(e.target.value)}
            warn={changesMade && !isNameValid}
          />
          <div>
            <Label htmlFor="link">
              <H3>Link</H3>
              {changesMade && !isLinkValid && <P warn>Link is not valid</P>}
            </Label>
            <Textarea
              warn={changesMade && !isLinkValid}
              id="link"
              data-testid="linkInput"
              onChange={(e) => setLink(e.target.value)}
              value={_link}
            />
          </div>
          <br />

          {success && <p>Update Successful</p>}

          <br />

          <BtnContainer>
            <Button
              disabled={!submitEnabled}
              data-testid="submit"
              as="input"
              type="submit"
              value="Save"
            />
            {initialSong && (
              <Button onClick={deleteSelf} color="salmon">
                Delete
              </Button>
            )}
          </BtnContainer>
        </>
      )}
    </Description>
  );
};

_EditableTile.propTypes = {
  name: PropTypes.string.isRequired,
  link: PropTypes.string.isRequired,

  /**
   * If the song resource is already in memory somewhere, this will prevent a
   * request from being made if it can optionally be passed as a prop.
   */
  initialSong: PropTypes.object,

  /* Should take two arguments: name and songId */
  onSave: PropTypes.func,
};

const mapStateToProps = (state) => ({
  token: state.auth.token,
});

export const EditableTile = connect(mapStateToProps)(_EditableTile);
