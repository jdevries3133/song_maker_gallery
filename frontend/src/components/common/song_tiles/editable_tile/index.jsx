export { EditableTile } from "./editable_tile";
