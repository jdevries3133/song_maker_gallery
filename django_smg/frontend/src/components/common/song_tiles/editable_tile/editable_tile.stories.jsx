import React from "react";

import { BackendDependentStory } from "Common/storybook_util";

import { EditableTile } from "./editable_tile";

export default {
  title: "Common/Song Tiles/Editable Tile",
  component: EditableTile,
};

const Template = (args) => (
  <BackendDependentStory>
    <EditableTile {...args} />
  </BackendDependentStory>
);

export const Default = Template.bind({});
Default.args = {
  name: "Frankie",
  link:
    "https://musiclab.chromeexperiments.com/Song-Maker/song/6594803161104384",
};
