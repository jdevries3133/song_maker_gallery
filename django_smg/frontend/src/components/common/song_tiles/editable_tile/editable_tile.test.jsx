import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

import { EditableTile } from "./editable_tile";

const onSave = jest.fn();

beforeEach(() => {
  render(
    <EditableTile
      name="Frank G."
      link="https://musiclab.chromeexperiments.com/Song-Maker/song/6594803161104384"
      onSave={onSave}
    />
  );
});

afterEach(() => onSave.mockClear());

describe("<EditableTile />", () => {
  it("matches snapshot", () => {
    expect(
      renderer
        .create(
          <EditableTile
            name="Frank G."
            link="https://musiclab.chromeexperiments.com/Song-Maker/song/6594803161104384"
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("doesn't show submit button without name", () => {
    fireEvent.change(screen.queryByTestId("linkInput"), {
      target: { value: "different" },
    });
    fireEvent.change(screen.queryByTestId("nameInput"), {
      target: { value: "" },
    });
    expect(screen.queryByText("Please enter your name.")).toBeVisible();
    expect(onSave).toHaveBeenCalledTimes(0);
  });

  it("calls passed func if data is valid", () => {
    fireEvent.change(screen.queryByTestId("linkInput"), {
      target: {
        value:
          "https://musiclab.chromeexperiments.com/Song-Maker/song/1234567812345678",
      },
    });
    fireEvent.click(screen.queryByTestId("submit"));
    expect(onSave).toHaveBeenCalledWith("Frank G.", "1234567812345678");
  });

  describe("link validation", () => {
    const changeTo = (link) => {
      fireEvent.change(screen.getByTestId("linkInput"), {
        target: { value: link },
      });
    };

    const linkInput = () => screen.getByTestId("linkInput");

    it("validates domain", async () => {
      expect(linkInput()).toHaveStyle({ border: "1px solid #ccc" });
      changeTo(
        "https://musiclb.chromeexperiments.com/Song-Maker/song/6594803161104384"
      );
      await waitFor(() => {
        expect(screen.getByText("Link is not valid")).toBeVisible();
        expect(linkInput()).toHaveStyle({ border: "2px solid #e41000" });
      });
    });
    it("validates path", async () => {
      expect(linkInput()).toHaveStyle({ border: "1px solid #ccc" });
      changeTo(
        "https://musiclab.chromeexperiments.com/Song-Maker/embed/6594803161104384"
      );
      await waitFor(() => {
        expect(screen.getByText("Link is not valid")).toBeVisible();
        expect(linkInput()).toHaveStyle({ border: "2px solid #e41000" });
      });
    });
    it("validates id", async () => {
      expect(linkInput()).toHaveStyle({ border: "1px solid #ccc" });
      changeTo(
        "https://musiclab.chromeexperiments.com/Song-Maker/embed/6594a03161104384"
      );
      await waitFor(() => {
        expect(screen.getByText("Link is not valid")).toBeVisible();
        expect(linkInput()).toHaveStyle({ border: "2px solid #e41000" });
      });
    });
    it("trims whitespace", async () => {
      expect(linkInput()).toHaveStyle({ border: "1px solid #ccc" });
      changeTo(
        "    https://musiclab.chromeexperiments.com/Song-Maker/song/6594803161104384    "
      );
      await waitFor(() => {
        expect(screen.queryByText("Link is not valid")).toBeNull();
        expect(linkInput()).toHaveStyle({ border: "1px solid #ccc" });
      });
    });
  });
});
