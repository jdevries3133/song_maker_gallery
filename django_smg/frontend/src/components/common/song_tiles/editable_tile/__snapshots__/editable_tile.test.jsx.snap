// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableTile /> matches snapshot 1`] = `
.c2 {
  display: block;
  font-size: 1.17em;
  margin-block-start: 1em;
  margin-block-end: 1em;
  margin-inline-start: 0px;
  margin-inline-end: 0px;
  font-weight: bold;
}

.c3 {
  max-width: 100%;
  font-size: 20px;
  padding: 10px;
  border-radius: 5px;
}

.c3[type="checkbox"] {
  width: 30px;
  height: 30px;
}

.c4 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 19px;
  line-height: 1.5;
  margin: 0px;
  width: 80%;
  box-shadow: 1px 1px 1px #999;
  border: 1px solid #ccc;
  border-radius: 5px;
}

.c0 {
  text-align: center;
  margin: 3vw;
  padding: 3vw;
  font-size: 18px;
  background: white;
  border-radius: 20px;
  box-shadow: 0px 3px 8px rgb(100,100,100);
}

.c1 {
  margin: 0;
  padding: 1.5vw;
}

@media (max-width:600px) {
  .c0 {
    font-size: 16px;
    line-height: 1.5rem;
  }
}

<form
  className="c0 c1"
  onSubmit={[Function]}
>
  <label
    className="c2"
    htmlFor="name"
  >
    Name
  </label>
  <input
    className="c3"
    data-testid="nameInput"
    id="name"
    onChange={[Function]}
    value="Frank G."
  />
  <div>
    <label
      htmlFor="link"
    >
      <h3
        className="styles__H3-kZrHQh"
      >
        Link
      </h3>
    </label>
    <textarea
      className="c4"
      data-testid="linkInput"
      id="link"
      onChange={[Function]}
      value="https://musiclab.chromeexperiments.com/Song-Maker/song/6594803161104384"
    />
  </div>
</form>
`;
